PSModule/AlmLabMgmtClient/SDK/Auth/RestAuthenticator.cs

```diff
@@ -0,0 +1,133 @@
+﻿using PSModule.AlmLabMgmtClient.SDK.Interface;
+using PSModule.AlmLabMgmtClient.SDK.Util;
+using System;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+
+namespace PSModule.AlmLabMgmtClient.SDK.Auth
+{
+    using C = Constants;
+    public class RestAuthenticator : IAuthenticator
+    {
+        private const string USERNAME = "Username";
+        private const string IS_AUTH_ENDPOINT = "rest/is-authenticated";
+        private const string AUTH_ENDPOINT = "authentication-point/authenticate";
+        private const string LOGOUT_ENDPOINT = "authentication-point/logout";
+        private const string SESSION_ENDPOINT = "rest/site-session";
+        private const string LOGGED_OUT_SUCCESSFULLY = "Logged Out Successfully.";
+        private const string CHECK_IF_AUTHENTICATED = "Check if is authenticated ...";
+        protected const string AUTHENTICATE_HEADER = "WWW-Authenticate";
+        protected const string AUTHENTICATION_INFO = "AuthenticationInfo";
+
+        public async Task<bool> Login(IClient client, string username, string password, string clientType)
+        {
+            try
+            {
+                bool isAuthenticated = await IsAuthenticated(client, username);
+                if (isAuthenticated)
+                    return true;
+
+                bool ok = await Authenticate(client, username, password);
+
+                if (ok)
+                    ok = await AppendQCSessionCookies(client, clientType);
+                return ok;
+            }
+            catch (ThreadInterruptedException)
+            {
+                throw;
+            }
+            catch (Exception e)
+            {
+                await client.Logger.LogError(e.Message);
+                return false;
+            }
+        }
+
+        private async Task<bool> AppendQCSessionCookies(IClient client, string clientType)
+        {
+            await client.Logger.LogInfo("Creating session...");
+            var headers = new WebHeaderCollection
+            {
+                { HttpRequestHeader.ContentType, C.APP_XML },
+                { HttpRequestHeader.Accept, C.APP_XML }
+            };
+
+            // issue a post request so that cookies relevant to the QC Session will be added to the RestClient
+            Response res = await client.HttpPost(
+                client.ServerUrl.AppendSuffix(SESSION_ENDPOINT), 
+                headers,
+                $"<session-parameters><client-type>{clientType}</client-type></session-parameters>");

```

Comment:
Is that possible the `clientType` may contain `<`, `>` or other invalid XML characters? If so, may need to use `CDATA` to make sure the string value shall not break the XML rules.

Reply:
Invalid values are allowed at least from Azure UI (in most cases this field is left empty). 
For this issue I would ask the QA to reproduce and open a defect, then we can fix and validate it specifically.
What do you think?

Reply:
Sure. No problem.